ocaml/typing/ctype.mli

```diff
@@ -404,6 +405,35 @@ val subtype: Env.t -> type_expr -> type_expr -> unit -> unit
            enforce and returns a function that enforces this
            constraints. *)
 
+(* This module allows a type to become "rigid": after unifying such a type,
+   we can check to see whether any of its variables were unified, issuing
+   an error in such a case. *)
+module Rigidify : sig
+  type t
+
+  type matches_result =
+    (* A variable with name [name] has been unified with a type [ty]. *)
+    | Unification_failure of
+        { name : string option
+        ; ty : type_expr }
+
+    (* A variable [ty] started with an [original_jkind] but now has a
+       [inferred_jkind]. *)
+    | Jkind_mismatch of
+        { original_jkind : jkind; inferred_jkind : jkind; ty : type_expr }
+
+    (* No problems *)
+    | All_good
+
+  (* Mark all the variables in the types given as rigid; these will be tracked
+     for unification. Every call to this function should be paired with a call
+     to [all_distinct_vars_with_original_jkinds]. *)
+  val rigidify_list : type_expr list -> t
+
+  (* Check that no variables in a [t] have actually been unified. *)
+  val all_distinct_vars_with_original_jkinds : Env.t -> t -> matches_result
```

Comment:
I think I looked into this at some point, but there's not always the right bracketing relationship in the source code between the two function calls, and so it would be invasive. I decided no to, in the end. (I think we should just Resolve and move on -- Resolve if you agree.)